### ISSUE TITLE
"Error: disk: duplicate name orphaned_disk_0", while disks are ignored.

### ISSUE THREAD

**Issue by FrederikNS** [terraform-providers/terraform-provider-vsphere#875]
We run a kubernetes cluster on CoreOS VMware VMs provisioned by Terraform. The VMs created correctly, and we installed Kubernetes subsequently. Kubernetes dynamically adds volumes to the nodes. Because of this we set terraform to ignore the volumes.

This worked fine for a while. However suddenly Terraform has started failing mentioning `Error: disk: duplicate name orphaned_disk_0`.

I suspect terraform is trying to disconnect multiple volumes at the same time. But as we have ignored changes to the disks, this shouldn't be a problem.

### Terraform Version

```
Terraform v0.12.12
+ provider.ignition v1.1.0
+ provider.local v1.4.0
+ provider.null v2.1.2
+ provider.vsphere v1.13.0
```

### vSphere Provider Version

```
.
├── module.coreos
│   └── provider.ignition ~> 1.1
├── module.kubespray
│   ├── provider.local ~> 1.3
│   └── provider.null ~> 2.1
└── module.vsphere
    └── provider.vsphere 1.13.0
```

### Affected Resource(s)

Please list the resources as a list, for example:
- `vsphere_virtual_machine`

### Terraform Configuration Files

```hcl
resource "vsphere_virtual_machine" "master" {
  count            = "${var.vm_master_count}"
  name             = "${var.vm_name_prefix}-master-${count.index}"
  resource_pool_id = "${vsphere_resource_pool.resource_pool.id}"
  datastore_id     = "${data.vsphere_datastore.datastore.id}"
  folder           = "${vsphere_folder.folder.path}"

  num_cpus         = "${var.vm_master_cpu}"
  memory           = "${var.vm_master_ram}"
  guest_id         = "${data.vsphere_virtual_machine.template.guest_id}"
  enable_disk_uuid = "true"

  network_interface {
    network_id   = "${data.vsphere_network.network.id}"
    adapter_type = "${data.vsphere_virtual_machine.template.network_interface_types[0]}"
  }

  disk {
    label            = "${var.vm_name_prefix}-master-${count.index}.vmdk"
    size             = "${data.vsphere_virtual_machine.template.disks.0.size}"
    eagerly_scrub    = "${data.vsphere_virtual_machine.template.disks.0.eagerly_scrub}"
    thin_provisioned = "${data.vsphere_virtual_machine.template.disks.0.thin_provisioned}"
  }

  disk {
    label            = "${var.vm_name_prefix}-master-${count.index}-docker.vmdk"
    size             = "${var.vm_master_docker_size}"
    eagerly_scrub    = "false"
    thin_provisioned = "true"
    unit_number      = "1"
  }

  clone {
    template_uuid = "${data.vsphere_virtual_machine.template.id}"
    linked_clone  = "${var.vm_linked_clone}"
  }

  vapp {
    properties = {
      "guestinfo.coreos.config.data" = var.vm_guestinfo
    }
  }

  lifecycle {
    ignore_changes = [
      disk
    ]
  }
}

# Create the Kubernetes worker VMs #
resource "vsphere_virtual_machine" "worker" {
  count            = "${var.vm_worker_count}"
  name             = "${var.vm_name_prefix}-worker-${count.index}"
  resource_pool_id = "${vsphere_resource_pool.resource_pool.id}"
  datastore_id     = "${data.vsphere_datastore.datastore.id}"
  folder           = "${vsphere_folder.folder.path}"

  num_cpus         = "${var.vm_worker_cpu}"
  memory           = "${var.vm_worker_ram}"
  guest_id         = "${data.vsphere_virtual_machine.template.guest_id}"
  enable_disk_uuid = "true"

  network_interface {
    network_id   = "${data.vsphere_network.network.id}"
    adapter_type = "${data.vsphere_virtual_machine.template.network_interface_types[0]}"
  }

  disk {
    label            = "${var.vm_name_prefix}-worker-${count.index}.vmdk"
    size             = "${data.vsphere_virtual_machine.template.disks.0.size}"
    eagerly_scrub    = "${data.vsphere_virtual_machine.template.disks.0.eagerly_scrub}"
    thin_provisioned = "${data.vsphere_virtual_machine.template.disks.0.thin_provisioned}"
  }

  disk {
    label            = "${var.vm_name_prefix}-worker-${count.index}-docker.vmdk"
    size             = "${var.vm_worker_docker_size}"
    eagerly_scrub    = "false"
    thin_provisioned = "true"
    unit_number      = "1"
  }

  clone {
    template_uuid = "${data.vsphere_virtual_machine.template.id}"
    linked_clone  = "${var.vm_linked_clone}"
  }

  vapp {
    properties = {
      "guestinfo.coreos.config.data" = var.vm_guestinfo
    }
  }

  lifecycle {
    ignore_changes = [
      disk
    ]
  }
}
```

### Debug Output

https://gist.github.com/FrederikNS/d82eac1d1c8344c09840b2ff32d407ca

### Expected Behavior

I expected Terraform to completely ignore dynamic volumes, due to the lifecycle ignoring the disk.

### Actual Behavior

Terraform complains that `Error: disk: duplicate name orphaned_disk_0`. 

I suspect that Terraform makes a sanity check on the volumes, where it plans to remove all the dynamic volumes from the nodes, and inadverdently creates a name conflict. Eventhough it should completely ignore the disks.

### Steps to Reproduce

1. Create a number of VMs in VMware vSphere using `terraform apply`.
2. Set the "ignore disks" lifecycle.
2. Add some volumes to the VMs.
3. Run `terraform plan` or `terraform apply`.
4. Receive an error about duplicate orphaned volumes.

### Important Factoids

The VMs run a Kubernetes cluster, which can dynamically add disks to the VMware VMs.

**Comment by branttaylor** [comment]
Seeing this issue in our environment as well. We also are using Terraform to provision Kubernetes nodes and therefore have the same pattern of disks moving between nodes fairly regularly (disks not created by Terraform).

**Comment by pvlltvk** [comment]
Having same issue with v0.12.16

**Comment by Fodoj** [comment]
Having the same issue on v0.12.18

**Comment by FrederikNS** [comment]
> @FrederikNS were you able to work around this? Or did you just go back to an older version of Terraform?

No I haven't been able to work around this issue... Currently I'm provisioning new nodes and changing old ones using the `-target` parameter. That mostly solves the issue for my use case, but all the following dependent bits are unable to run at all.

**Comment by AlekseyUsov** [comment]
> > @FrederikNS were you able to work around this? Or did you just go back to an older version of Terraform?
> 
> No I haven't been able to work around this issue... Currently I'm provisioning new nodes and changing old ones using the `-target` parameter. That mostly solves the issue for my use case, but all the following dependent bits are unable to run at all.

Hit this issue today with OpenShift 3.11. Was able to get around it by draining each node before applying terraform configuration:
`oc adm drain <node> --delete-local-data --ignore-daemonsets --force`
There is a similar k8s command - `kubectl drain`.

**Comment by gricuk** [comment]
I have the same issue, does anyone have any updates?

**Comment by ajchiarello** [comment]
Is there any work being done to resolve this? It's becoming a real problem in my environment.

**Comment by Jetersen** [comment]
I was diving into the code but I have yet to any experience contributing to the project nor do I know how best to solve the issue.

My initial guess is to simply ignore orphaned disks but no actual code and no way of testing it :)

**Comment by Fodoj** [comment]
Check if you can be hitting this one https://github.com/terraform-providers/terraform-provider-vsphere/issues/977

**Comment by Jetersen** [comment]
@Fodoj we hit that one yesterday while demonstrating the bug

**Comment by Fodoj** [comment]
Note the #977 happens with any disk that has not exact number of gbs in size, like thin provisioned ones (vcenter will report something like 123.45454, Terraform will understand 123 and then you get this error).

**Comment by Fodoj** [comment]
I was totally wrong, this has _nothing_ to do with thin provisioned volumes. I did find the root cause though, see #1005

**Comment by Jetersen** [comment]
Fixed in #1005 which was merged 🎉
